Host-Linux Nginx: Add RHEL and SLES distros (#27774)
dotnet/AspNetCore.Docs: aspnetcore/host-and-deploy/linux-nginx.md

---
title: Host ASP.NET Core on Linux with Nginx
author: rick-anderson

> description: Learn how to set up Nginx as a reverse proxy on Ubuntu 20.04 to forward HTTP traffic to an ASP.NET Core web app running on Kestrel.

Reviewer: Should we list the distros here too? RHEL/SUSE.

Author: Yes!  Fixed meta description, thanks.  I also took out the version in the meta data description since several versions apply.

description: Learn how to set up Nginx as a reverse proxy on an Ubuntu 20.04 VM to forward HTTP traffic to an ASP.NET Core web app running on Kestrel.
monikerRange: '>= aspnetcore-3.1'
ms.author: riande
ms.custom: mvc, engagement-fy23
ms.date: 01/08/2023
uid: host-and-deploy/linux-nginx
---
# Host ASP.NET Core on Linux with Nginx

By [Sourabh Shirhatti](https://twitter.com/sshirhatti)

:::moniker range=">= aspnetcore-6.0"

> This guide explains setting up a production-ready ASP.NET Core environment on a VM for Ubuntu, Red Hat Enterprise (RHEL) and SUSE Linux Enterprise Server.

Reviewer: Missing comma
```suggestion
This guide explains setting up a production-ready ASP.NET Core environment on a VM for Ubuntu, Red Hat Enterprise (RHEL), and SUSE Linux Enterprise Server.
```

## Prerequisites

# [Ubuntu](#tab/linux-ubuntu)

> * Access to an Ubuntu 20.04 VM with a standard user account with sudo privilege. These instructions likely work with newer versions of Ubuntu, but the instructions haven't been tested with newer versions.

Reviewer 2: ```suggestion
* Access to an Ubuntu 20.04 VM with a standard user account with sudo privilege. These instructions haven't been tested with the latest Ubuntu versions,  but they should work.
```

* The latest stable [.NET runtime installed](/dotnet/core/install/linux) on the server.
* An existing ASP.NET Core app.

# [Red Hat Enterprise Linux](#tab/linux-rhel)

* Access to a Red Hat Enterprise (RHEL) 8.0 or later VM with a standard user account with sudo privilege.
* The latest stable [.NET runtime installed](/dotnet/core/install/linux) on the server.
* An existing ASP.NET Core app.

# [SUSE Linux Enterprise Server](#tab/linux-sles)

> * Access to an SLES 12 VM with a standard user account with sudo privilege.

Reviewer 3: SLES 12 and 15 are supported.

### Configure Nginx

# [Ubuntu](#tab/linux-ubuntu)

To configure Nginx as a reverse proxy to forward HTTP requests to your ASP.NET Core app, modify `/etc/nginx/sites-available/default`. Open it in a text editor, and replace the contents with the following snippet:

# [Red Hat Enterprise Linux](#tab/linux-rhel)

> To configure Nginx as a reverse proxy to forward HTTP requests to your ASP.NET Core app, modify `/etc/nginx.conf`. Open it in a text editor, and replace the `server{}` code block with the following snippet:

Reviewer 2: ```suggestion
To configure Nginx as a reverse proxy to forward HTTP requests to an ASP.NET Core app, modify `/etc/nginx.conf`. Replace the `server{}` code block with the following snippet:
```

# [SUSE Linux Enterprise Server](#tab/linux-sles)

> To configure Nginx as a reverse proxy to forward HTTP requests to your ASP.NET Core app, modify `/etc/nginx.conf`. Open it in a text editor, and replace the `server{}` code block with the following snippet:

Reviewer 2: same as previous change

1. Navigate to the app's directory.
1. Run the app: `dotnet <app_assembly.dll>`, where `app_assembly.dll` is the assembly file name of the app.

# [Ubuntu](#tab/linux-ubuntu)

If the app runs on the server but fails to respond over the Internet, check the server's firewall and confirm port 80 is open. If using an Azure Ubuntu VM, add a Network Security Group (NSG) rule that enables inbound port 80 traffic. There's no need to enable an outbound port 80 rule, as the outbound traffic is automatically granted when the inbound rule is enabled.

# [Red Hat Enterprise Linux](#tab/linux-rhel)

> If the app runs on the server but fails to respond over the Internet, check the server's firewall and confirm port 80 is open.

Reviewer: I realize you're copy/pasting the original and tweaking it, but internet shouldn't be capitalized. There are a few more places that need this change, I didn't point them out. I wonder if Acrolinx was ever run on the original article?
```suggestion
If the app runs on the server but fails to respond over the internet, check the server's firewall and confirm port 80 is open.
```

# [SUSE Linux Enterprise Server](#tab/linux-sles)

If the app runs on the server but fails to respond over the Internet, check the server's firewall and confirm port 80 is open.

---

When done testing the app, shut down the app with <kbd>Ctrl</kbd>+<kbd>C</kbd> (Windows) or <kbd>⌘</kbd>+<kbd>C</kbd> (macOS) at the command prompt.

## Monitor the app

> The server is set up to forward requests made to `http://<serveraddress>:80` on to the ASP.NET Core app running on Kestrel at `http://127.0.0.1:5000`. However, Nginx isn't set up to manage the Kestrel process. `systemd` can be used to create a service file to start and monitor the underlying web app. `systemd` is an init system that provides many powerful features for starting, stopping, and managing processes.

Reviewer 2: make sysemd a link

Edit `src/http/ngx_http_header_filter_module.c`:

> ```text

Reviewer 2: that's C, not text

sudo ufw enable
```

# [Red Hat Enterprise Linux](#tab/linux-rhel)

> [!WARNING]

> > A firewall will prevent access to the whole system if not configured correctly. Failure to specify the correct SSH port will effectively lock you out of the system if you are using SSH to connect to it. The default port is 22. For more information, see the [introduction to ufw](https://help.ubuntu.com/community/UFW).

Reviewer 2: No future case
Remove `you` unless necessary

Install `ufw` and configure it to allow traffic on any ports needed.

```bash
sudo yum -y install ufw

sudo ufw allow 22/tcp
sudo ufw allow 80/tcp
sudo ufw allow 443/tcp

sudo ufw enable
```

# [SUSE Linux Enterprise Server](#tab/linux-sles)

> [!WARNING]

> > A firewall will prevent access to the whole system if not configured correctly. Failure to specify the correct SSH port will effectively lock you out of the system if you are using SSH to connect to it. The default port is 22. For more information, see the [introduction to ufw](https://help.ubuntu.com/community/UFW).

Reviewer 2: same

[!code-nginx[](linux-nginx/proxy.conf)]

# [Ubuntu](#tab/linux-ubuntu)

**Replace** the contents of the */etc/nginx/nginx.conf* configuration file with the following file. The example contains both `http` and `server` sections in one configuration file.

> [!code-nginx[](linux-nginx/nginx.conf)]

Reviewer: This one needs to be deleted since it's been moved to be outside the tabs. Otherwise, the Ubuntu tab is showing two of the same code blocks.

Author: Ouch!  Good catch!

# [Red Hat Enterprise Linux](#tab/linux-rhel)

> Modify `/etc/nginx/nginx.conf`. Open it in a text editor, and replace the `http{}` and `server{}` code blocks with contents with the following snippet:

Reviewer 2: same

# [SUSE Linux Enterprise Server](#tab/linux-sles)

> Modify `/etc/nginx/nginx.conf`. Open it in a text editor, and replace the `http{}` and `server{}` code blocks with contents with the following snippet:

Reviewer 2: same

---

> [!code-nginx[](linux-nginx/nginx.conf)]

Reviewer: I'm not sure why these are includes instead of snippets?

Author: I don't know the original author's intent.  Showing it is a .conf example in its entirety available perhaps?
